File: templates/python/permutation_null.py
"""permutation_null: 置换检验零分布（null 直方图 + 观测统计量 + 双侧 p 值阴影）."""
import sys, os; sys.path.insert(0, os.path.join(os.path.dirname(os.path.abspath(__file__)), '..', '..', '_utils', 'python'))
import numpy as np
import matplotlib.pyplot as plt
from theme import apply_theme
from palette import cycle

def make_figure(x=None, y=None, n_perm=5000, title='Permutation test null distribution'):
    apply_theme()
    rng = np.random.default_rng(6)
    if x is None:
        x = rng.normal(0.0, 1.0, 40)
        y = rng.normal(0.8, 1.0, 35)
    obs = y.mean() - x.mean()
    pooled = np.concatenate([x, y])
    nx = len(x)
    perm = rng.permuted(np.tile(pooled, (n_perm, 1)), axis=1)
    null = perm[:, nx:].mean(axis=1) - perm[:, :nx].mean(axis=1)
    p = (np.abs(null) >= abs(obs)).mean()
    fig, ax = plt.subplots()
    counts, edges, patches = ax.hist(null, bins=50, color=cycle(0), alpha=0.7,
                                     edgecolor='w', linewidth=0.3)
    for e, patch in zip(edges[:-1], patches):
        if abs(e + (edges[1] - edges[0]) / 2) >= abs(obs):
            patch.set_facecolor(cycle(1)); patch.set_alpha(0.9)
    ax.axvline(obs, color='k', linewidth=1.4,
               label=f'observed diff = {obs:.2f}')
    ax.axvline(-obs, color='k', linewidth=1.0, linestyle=':')
    ax.text(0.02, 0.95, f'two-sided p = {p:.4f}\n({n_perm} permutations)',
            transform=ax.transAxes, va='top', fontsize=8)
    ax.set_xlabel('mean difference under null')
    ax.set_ylabel('count')
    ax.set_title(title)
    ax.legend(loc='upper right')
    ax.grid(True, axis='y', linestyle=':', alpha=0.5)
    fig.tight_layout()
    return fig

if __name__ == '__main__':
    fig = make_figure(); fig.savefig(__file__.replace('.py', '.png'), dpi=150)

File: templates/matlab/permutation_null.m
function fig = permutation_null()
    addpath(fullfile(fileparts(mfilename('fullpath')),'..','..','_utils','matlab'));
    apply_theme(); rng(6);
    n_perm = 5000;
    x = randn(1, 40);
    y = 0.8 + randn(1, 35);
    obs = mean(y) - mean(x);
    pooled = [x y];
    nx = numel(x); N = numel(pooled);
    null_v = zeros(1, n_perm);
    for p = 1:n_perm
        idx = randperm(N);
        null_v(p) = mean(pooled(idx(nx+1:end))) - mean(pooled(idx(1:nx)));
    end
    pval = mean(abs(null_v) >= abs(obs));
    [counts, edges] = histcounts(null_v, 50);     % manual histogram, no toolbox
    centers = (edges(1:end-1) + edges(2:end))/2;
    tail = abs(centers) >= abs(obs);
    fig = figure; hold on;
    bar(centers, counts.*(~tail), 1, 'FaceColor', palette('cat',1), ...
        'FaceAlpha', 0.7, 'EdgeColor', 'w', 'LineWidth', 0.3);
    bar(centers, counts.*tail, 1, 'FaceColor', palette('cat',2), ...
        'FaceAlpha', 0.9, 'EdgeColor', 'w', 'LineWidth', 0.3);
    ym = max(counts)*1.05;
    hob = plot([obs obs], [0 ym], 'k-', 'LineWidth', 1.4);
    plot([-obs -obs], [0 ym], 'k:', 'LineWidth', 1.0);
    text(0.02, 0.95, sprintf('two-sided p = %.4f\n(%d permutations)', pval, n_perm), ...
         'Units', 'normalized', 'VerticalAlignment', 'top', 'FontSize', 8);
    ylim([0 ym]);
    xlabel('mean difference under null'); ylabel('count');
    title('Permutation test null distribution');
    legend(hob, {sprintf('observed diff = %.2f', obs)}, 'Location', 'northeast');
    grid on; set(gca, 'XGrid', 'off');
end
